File: src/toonConverter.ts
/**
 * TOON (Token Optimized Object Notation) Converter
 * Converts JSON to a CSV-like token-efficient format
 *
 * Format:
 * - Simple properties: key,value
 * - Arrays of objects: arrayName[length]{key1,key2,...}
 *   followed by rows of values
 * - Nested objects: expanded with dot notation or inline
 */

import type { ConversionMetrics } from './types';

export function jsonToToon(jsonString: string): string {
    try {
        const obj = JSON.parse(jsonString);
        return convertToToon(obj, '');
    } catch (error) {
        throw new Error(`Invalid JSON: ${(error as Error).message}`);
    }
}

function convertToToon(obj: any, prefix: string = ''): string {
    if (obj === null || obj === undefined) {
        return '';
    }

    const lines: string[] = [];

    for (const [key, value] of Object.entries(obj)) {
        const fullKey = prefix ? `${prefix}.${key}` : key;

        if (value === null || value === undefined) {
            lines.push(`${fullKey},null`);
        } else if (typeof value === 'boolean') {
            lines.push(`${fullKey},${value}`);
        } else if (typeof value === 'number') {
            lines.push(`${fullKey},${value}`);
        } else if (typeof value === 'string') {
            const escapedValue = escapeValue(value);
            lines.push(`${fullKey},${escapedValue}`);
        } else if (Array.isArray(value)) {
            if (value.length === 0) {
                lines.push(`${fullKey}[0]`);
            } else if (isArrayOfObjects(value)) {
                lines.push(formatObjectArray(fullKey, value));
            } else if (isArrayOfPrimitives(value)) {
                const values = value.map(v =>
                    typeof v === 'string' ? escapeValue(v) : String(v)
                ).join(',');
                lines.push(`${fullKey}[${value.length}],${values}`);
            } else {
                value.forEach((item, index) => {
                    if (typeof item === 'object' && item !== null) {
                        lines.push(convertToToon(item, `${fullKey}[${index}]`));
                    } else {
                        const val = typeof item === 'string' ? escapeValue(item) : String(item);
                        lines.push(`${fullKey}[${index}],${val}`);
                    }
                });
            }
        } else if (typeof value === 'object') {
            lines.push(convertToToon(value, fullKey));
        }
    }

    return lines.filter(line => line).join('\n');
}

function formatObjectArray(key: string, array: any[]): string {
    if (array.length === 0) return `${key}[0]`;

    const allKeys = new Set<string>();
    array.forEach(obj => {
        Object.keys(obj).forEach(k => allKeys.add(k));
    });
    const keys = Array.from(allKeys);

    const header = `${key}[${array.length}]{${keys.join(',')}}`;

    const rows = array.map(obj => {
        return keys.map(k => {
            const value = obj[k];
            if (value === null || value === undefined) return 'null';
            if (typeof value === 'boolean') return value;
            if (typeof value === 'number') return value;
            if (typeof value === 'string') return escapeValue(value);
            if (typeof value === 'object') return JSON.stringify(value);
            return String(value);
        }).join(',');
    });

    return [header, ...rows.map(row => '  ' + row)].join('\n');
}

function isArrayOfObjects(arr: any[]): boolean {
    if (arr.length === 0) return false;
    return arr.every(item =>
        item !== null &&
        typeof item === 'object' &&
        !Array.isArray(item)
    );
}

function isArrayOfPrimitives(arr: any[]): boolean {
    if (arr.length === 0) return false;
    return arr.every(item =>
        item === null ||
        typeof item === 'string' ||
        typeof item === 'number' ||
        typeof item === 'boolean'
    );
}

function escapeValue(value: string): string {
    if (value.includes(',') || value.includes('\n') || value.includes('"')) {
        return '"' + value.replace(/"/g, '""') + '"';
    }
    return value;
}

export function estimateTokens(text: string): number {
    if (!text) return 0;

    const charCount = text.length;
    const wordCount = text.split(/\s+/).filter(w => w.length > 0).length;
    const specialChars = (text.match(/[{}[\]:,\n]/g) || []).length;

    const charBasedEstimate = charCount / 4;
    const wordBasedEstimate = wordCount + specialChars * 0.5;

    return Math.round((charBasedEstimate * 0.6 + wordBasedEstimate * 0.4));
}

export function calculateMetrics(jsonString: string, toonString: string): ConversionMetrics {
    const jsonTokens = estimateTokens(jsonString);
    const toonTokens = estimateTokens(toonString);
    const tokensSaved = jsonTokens - toonTokens;
    const reductionPercent = jsonTokens > 0
        ? ((tokensSaved / jsonTokens) * 100).toFixed(1)
        : '0';

    return {
        jsonTokens,
        toonTokens,
        tokensSaved,
        reductionPercent
    };
}

export function validateTokenLimit(text: string, maxTokens: number = 1000000): number {
    const tokens = estimateTokens(text);
    if (tokens > maxTokens) {
        throw new Error(`Input exceeds maximum token limit of ${maxTokens.toLocaleString()}. Current: ${tokens.toLocaleString()}`);
    }
    return tokens;
}

File: backend/toon_converter.py
"""
TOON (Token Optimized Object Notation) Converter
Converts JSON to a CSV-like token-efficient format
"""

import json
import re
from typing import Any, Dict, List, Set


def json_to_toon(json_string: str) -> str:
    """Convert JSON string to TOON format"""
    try:
        obj = json.loads(json_string)
        return convert_to_toon(obj, '')
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON: {str(e)}")


def convert_to_toon(obj: Any, prefix: str = '') -> str:
    """Recursively convert object to TOON format"""
    if obj is None:
        return ''

    lines: List[str] = []

    # Handle root-level arrays (from CSV/XLSX uploads)
    if isinstance(obj, list):
        if len(obj) == 0:
            return 'data[0]'
        elif is_array_of_objects(obj):
            return format_object_array('data', obj)
        elif is_array_of_primitives(obj):
            values = ','.join(
                escape_value(v) if isinstance(v, str) else str(v)
                for v in obj
            )
            return f"data[{len(obj)}],{values}"
        else:
            # Mixed array
            for index, item in enumerate(obj):
                if isinstance(item, dict):
                    lines.append(convert_to_toon(item, f"data[{index}]"))
                else:
                    val = escape_value(item) if isinstance(item, str) else str(item)
                    lines.append(f"data[{index}],{val}")
            return '\n'.join(filter(None, lines))

    if not isinstance(obj, dict):
        return ''

    for key, value in obj.items():
        full_key = f"{prefix}.{key}" if prefix else key

        if value is None:
            lines.append(f"{full_key},null")
        elif isinstance(value, bool):
            lines.append(f"{full_key},{str(value).lower()}")
        elif isinstance(value, (int, float)):
            lines.append(f"{full_key},{value}")
        elif isinstance(value, str):
            escaped_value = escape_value(value)
            lines.append(f"{full_key},{escaped_value}")
        elif isinstance(value, list):
            if len(value) == 0:
                lines.append(f"{full_key}[0]")
            elif is_array_of_objects(value):
                lines.append(format_object_array(full_key, value))
            elif is_array_of_primitives(value):
                values = ','.join(
                    escape_value(v) if isinstance(v, str) else str(v)
                    for v in value
                )
                lines.append(f"{full_key}[{len(value)}],{values}")
            else:
                # Mixed array - indexed format
                for index, item in enumerate(value):
                    if isinstance(item, dict):
                        lines.append(convert_to_toon(item, f"{full_key}[{index}]"))
                    else:
                        val = escape_value(item) if isinstance(item, str) else str(item)
                        lines.append(f"{full_key}[{index}],{val}")
        elif isinstance(value, dict):
            lines.append(convert_to_toon(value, full_key))

    return '\n'.join(filter(None, lines))


def format_object_array(key: str, array: List[Dict]) -> str:
    """Format array of objects in compact table format"""
    if len(array) == 0:
        return f"{key}[0]"

    # Get all unique keys
    all_keys: Set[str] = set()
    for obj in array:
        all_keys.update(obj.keys())
    keys = list(all_keys)

    # Header: arrayName[length]{key1,key2,...}
    header = f"{key}[{len(array)}]{{{','.join(keys)}}}"

    # Data rows
    rows = []
    for obj in array:
        row_values = []
        for k in keys:
            value = obj.get(k)
            if value is None:
                row_values.append('null')
            elif isinstance(value, bool):
                row_values.append(str(value).lower())
            elif isinstance(value, (int, float)):
                row_values.append(str(value))
            elif isinstance(value, str):
                row_values.append(escape_value(value))
            elif isinstance(value, (dict, list)):
                row_values.append(json.dumps(value))
            else:
                row_values.append(str(value))
        rows.append('  ' + ','.join(row_values))

    return '\n'.join([header] + rows)


def is_array_of_objects(arr: List) -> bool:
    """Check if array contains only objects"""
    if len(arr) == 0:
        return False
    return all(isinstance(item, dict) for item in arr)


def is_array_of_primitives(arr: List) -> bool:
    """Check if array contains only primitives"""
    if len(arr) == 0:
        return False
    return all(
        item is None or isinstance(item, (str, int, float, bool))
        for item in arr
    )


def escape_value(value: str) -> str:
    """Escape values containing commas, newlines, or quotes"""
    if ',' in value or '\n' in value or '"' in value:
        return '"' + value.replace('"', '""') + '"'
    return value


def estimate_tokens(text: str) -> int:
    """Estimate token count for text"""
    if not text:
        return 0

    char_count = len(text)
    word_count = len([w for w in text.split() if w])
    special_chars = len(re.findall(r'[{}\[\]:,\n]', text))

    char_based_estimate = char_count / 4
    word_based_estimate = word_count + special_chars * 0.5

    return round(char_based_estimate * 0.6 + word_based_estimate * 0.4)


def calculate_metrics(json_string: str, toon_string: str) -> Dict[str, Any]:
    """Calculate conversion metrics"""
    json_tokens = estimate_tokens(json_string)
    toon_tokens = estimate_tokens(toon_string)
    tokens_saved = json_tokens - toon_tokens
    reduction_percent = (
        f"{(tokens_saved / json_tokens * 100):.1f}"
        if json_tokens > 0
        else "0"
    )

    return {
        "jsonTokens": json_tokens,
        "toonTokens": toon_tokens,
        "tokensSaved": tokens_saved,
        "reductionPercent": reduction_percent
    }
